### PR TITLE
Feature emissive and ao textures

### PR DESCRIPTION
**Pull request by LaisoEmilio**
Added gbuffer for emissive and ao textures. Also added support for emissive and ao textures to the hybrid/raytracing/pathtracing pipelines.

### CODE REVIEW

File: resources/shaders/deferred_composition.hlsl [hunk 1]
```diff
@@ -112,7 +123,7 @@ void main_cs(int3 dispatch_thread_id : SV_DispatchThreadID)
 			is_hybrid);
 
 		// Shade pixel
-		retval = shade_pixel(pos, V, albedo, metallic, roughness, normal, irradiance, ao, reflection, sampled_brdf, shadow_factor);
+		retval = shade_pixel(pos, V, albedo, metallic, roughness, normal, irradiance, ao, reflection, sampled_brdf, shadow_factor) + emissive;
```
Comment (VZout) [inline resources/shaders/deferred_composition.hlsl]: Do this inside shade pixel instead.

File: resources/shaders/gi_util.hlsl [hunk 2]
```diff
@@ -1,170 +1,170 @@
...
+		float D = D_GGX(NdotH, max(0.05, roughness)); 
+		// G = Geometric shadowing term (Microfacets shadowing)
+		float G = G_SchlicksmithGGX(NdotL, NdotV, max(0.05, roughness));
+		// F = Fresnel factor (Reflectance depending on angle of incidence)
+		float3 F = F_Schlick(NdotH, metal, albedo);
+ 
+		float3 spec = (D * F * G) / ((4.0 * NdotL * NdotV + 0.001f));
+		float  ggx_probability = D * NdotH / (4 * LdotH);
+
+		return NdotL * irradiance * spec / (ggx_probability * (1.0f - diffuse_probability));
```
Comment (VZout) [inline resources/shaders/gi_util.hlsl]: Hard to diff but didn't add AO at all?

File: resources/shaders/path_tracer.hlsl [hunk 3]
```diff
@@ -188,21 +191,25 @@ void ReflectionHit(inout HitInfo payload, in MyAttributes attr)
...
+		g_textures[material.ao_id],
 		mip_level,
 		s0,
 		uv);
 
 	float3 albedo = output_data.albedo;
 	float roughness = output_data.roughness;
 	float metal = output_data.metallic;
-	
+	float3 emissive = pow(output_data.emissive, 2.2f);
```
Comment (VZout) [inline resources/shaders/path_tracer.hlsl]: pow should probably be done inside the material_util file.

File: resources/shaders/path_tracer.hlsl [hunk 4]
```diff
@@ -222,11 +229,13 @@ void ReflectionHit(inout HitInfo payload, in MyAttributes attr)
...
 }
 
 //Reflection skybox
 [shader("miss")]
 void ReflectionMiss(inout HitInfo payload)
 {
-	payload.color = skybox.SampleLevel(s0, SampleSphericalMap(WorldRayDirection()), 0);
-}
+	//payload.color = skybox.SampleLevel(s0, SampleSphericalMap(WorldRayDirection()), 0);
+	payload.color = 0;
```
Comment (VZout) [inline resources/shaders/path_tracer.hlsl]: Revert this.

File: resources/shaders/post_processing.hlsl [hunk 5]
```diff
@@ -26,7 +26,7 @@ void main(uint3 DTid : SV_DispatchThreadID)
 	color = SampleFXAA(input, s0, DTid.xy + 0.5f, resolution);
 	//uv = ZoomUV(uv, 0.75);
 	//float3 color = input.SampleLevel(s0, BarrelDistortUV(uv, 2), 0);
-	//float3 color = ChromaticAberrationV2(input, s0, uv, 0.2, 0.96f).rgb;
+	//color = ChromaticAberrationV2(input, s0, uv, 0.2, 0.96f).rgb;
```
Comment (VZout) [inline resources/shaders/post_processing.hlsl]: Revert this.

File: resources/shaders/deferred_composition.hlsl (hunk 1 above)
Comment (LaisoEmilio) [reply resources/shaders/deferred_composition.hlsl]: Done 👍

File: resources/shaders/path_tracer.hlsl (hunk 4 above)
Comment (LaisoEmilio) [reply resources/shaders/path_tracer.hlsl]: Oooopsie

File: tests/demo/demo.cpp [hunk 6]
```diff
@@ -134,7 +134,7 @@ int WispEntry()
 
 			memcpy(indices.data(), model->m_meshes[0]->m_indices.data(), indices.size() * sizeof(std::uint32_t));
 
-			SCENE::resources::model_pool->EditMesh<wr::VertexColor, std::uint32_t>(SCENE::resources::test_model->m_meshes[0].first, vertices, indices);
+			//SCENE::resources::model_pool->EditMesh<wr::VertexColor, std::uint32_t>(SCENE::resources::test_model->m_meshes[0].first, vertices, indices);
```
[the remark this answers is not in the document]
Comment (LaisoEmilio) [reply tests/demo/demo.cpp]: Done, removed the other one as well

File: tests/demo/engine_interface.hpp [hunk 7]
```diff
@@ -7,6 +7,7 @@
 #include "d3d12/d3d12_renderer.hpp"
 #include "imgui/ImGuizmo.h"
 #include "demo_frame_graphs.hpp"
+#include "scene_emibl.hpp"
```
[the remark this answers is not in the document]
Comment (LaisoEmilio) [reply tests/demo/engine_interface.hpp]: Just and hardcoded test, my bad... left that in

File: resources/shaders/gi_util.hlsl (hunk 2 above)
Comment (LaisoEmilio) [reply resources/shaders/gi_util.hlsl]: Done

File: resources/shaders/path_tracer.hlsl (hunk 3 above)
Comment (LaisoEmilio) [reply resources/shaders/path_tracer.hlsl]: done

**Comment by LaisoEmilio** [comment]
Acted on the comments 👍

File: tests/demo/scene_emibl.hpp [hunk 8]
```diff
@@ -15,6 +15,8 @@ namespace emibl_scene
 		// RESOURCES
 		//////////////////////////
 
+		static float emissive_multiplier = 1.0f;
```
Comment (VZout) [inline tests/demo/scene_emibl.hpp]: unused

File: resources/shaders/path_tracer.hlsl [hunk 9]
```diff
@@ -21,15 +21,18 @@ struct Material
 	float normal_id;
 	float roughness_id;
 	float metalicness_id;
+	float emissive_id;
+	float ao_id;
```
Comment (Nielsbishere) [inline resources/shaders/path_tracer.hlsl]: Please use uint instead of float

File: resources/shaders/raytracing.hlsl [hunk 10]
```diff
@@ -22,15 +22,18 @@ struct Material
...
 	MaterialData data;
 };
 
 struct Offset
 {
-    float material_idx;
-    float idx_offset;
-    float vertex_offset;
+	float material_idx;
+	float idx_offset;
```
Comment (Nielsbishere) [inline resources/shaders/raytracing.hlsl]: Please use uint instead of float

File: resources/shaders/rt_hybrid.hlsl [hunk 11]
```diff
@@ -21,6 +21,9 @@ struct Material
 	float normal_id;
 	float roughness_id;
 	float metalicness_id;
+	float emissive_id;
```
Comment (Nielsbishere) [inline resources/shaders/rt_hybrid.hlsl]: Don't use float pls

File: src/d3d12/d3d12_renderer.hpp [hunk 12]
```diff
@@ -69,6 +69,9 @@ namespace wr
 			float normal_id;
 			float roughness_id;
 			float metallicness_id;
+			float emissive_id;
```
Comment (Nielsbishere) [inline src/d3d12/d3d12_renderer.hpp]: Float used instead of uint

**Pull request merged**